static/app/utils/discover/genericDiscoverQuery.tsx

```diff
@@ -423,10 +423,14 @@ export function useGenericDiscoverQuery<T, P>(props: Props<T, P>) {
   return useQuery<T, QueryError>(
     [route, apiPayload],
     async () => {
-      const [resp] = await doDiscoverQuery<T>(api, url, apiPayload, {
+      const [data, , resp] = await doDiscoverQuery<T>(api, url, apiPayload, {
         queryBatching: props.queryBatching,
       });
-      return resp;
+      const pageLinks = resp?.getResponseHeader('Link');
+      if (!pageLinks) {
+        return data;
+      }
+      return {pageLinks, ...data};
```

Comment:
I'm a little scared of this. Injecting stuff into people's `data` will make them confused! Since `useDiscoverQuery` is only used in ~16 places, what do you think about changing the API a bit?

I think if `useGenericDiscoverQuery` returned `[data, statusText, response]` that'd be the most generic version, just the API result as-is. Then, `useDiscoverQuery` could pass that through, and anything that uses `useDiscoverQuery` would change from `const data = useDis...` to `const [data] = useDis...` and those places would also have access to all those headers if they needed it.

In `useSpansQuery` you could choose to manually pull out the `pageLinks` header and return it alongside `data` and `isLoading` as you already do

I think that might be better because it keeps generic things generic, and the response format is more predictable in more places

Reply:
@gggritso interesting suggestion and thanks for the comment!
If I understand this correctly (let me know if you think i'm not!), i'm not sure about this solution. 

> I think if `useGenericDiscoverQuery` returned `[data, statusText, response]` that'd be the most generic version, just the API result as-is. Then, `useDiscoverQuery` could pass that through, and anything that uses `useDiscoverQuery` would change from `const data = useDis...` to `const [data] = useDis...` and those places would also have access to all those headers if they needed it.

So rn `useGenericDiscoverQuery` returns a `useQueryResult`, ie {data, isLoading, isRefetching, ...},
1. If it returned `[data, statusText, response]` we would lose important info such as isLoading.
2. If you meant the query function returning `[data, statusText, response]` then we run into the same problem where we are injecting information into people's data. as you'll have
```
const {data} = useGenericDiscoverQuery
// data = [data, statusText, response]
```


Maybe a better solution would be to add an additonal property to the `useQueryResult`. THis way we retain all properties of `useQueryResult`, we don't inject anything into the data, and we are able to get the full response and status code directly keeping it all generic.

Something like this
```
export function useGenericDiscoverQuery<T, P>(props: Props<T, P>) {
  const api = useApi();
  const {orgSlug, route, options} = props;
  const url = `/organizations/${orgSlug}/${route}/`;
  const apiPayload = getPayload<T, P>(props);

  const result = useQuery<
    [T, string | undefined, ResponseMeta<T> | undefined],
    QueryError
  >(
    [route, apiPayload],
    () =>
      doDiscoverQuery<T>(api, url, apiPayload, {
        queryBatching: props.queryBatching,
      }),
    options
  );
  
  // this retains all properties of `useQuery` and allows you to get the statusCode and the full response
  return {
    ...result,
    data: result.data?.[0],
    statusCode: result.data?.[1],
    resp: result.data?.[2],
  };
}
```

And then `useDiscoverQuery` can return `useQueryResult & {pageLinks, response, statusCode}`. pageLinks is such a common thing, that I think it warrants its own property, rather then retrieving the header every single time.

Reply:
Yep, that sounds good to me!

Reply:
@gggritso I made the changes! If u could have another look when you have a chance that would be great!